File: 07/solution.ml
open Base
open Stdio

let read_input () = In_channel.read_lines "07/input.txt"

let neighbors grid (x, y) =
  let rows = Array.length grid in
  let cols = Array.length grid.(0) in
  let cell =
    if y >= 0 && y < rows && x >= 0 && x < cols then Some grid.(y).(x) else None
  in
  let deltas =
    match cell with
    | Some '.' | Some 'S' -> [ (0, 1) ]
    | Some '^' -> [ (1, 1); (-1, 1) ]
    | _ -> []
  in
  deltas
  |> List.map ~f:(fun (dx, dy) -> (x + dx, y + dy))
  |> List.filter ~f:(fun (nx, ny) ->
      nx >= 0 && nx < cols && ny >= 0 && ny < rows)

let bfs grid start =
  let to_key (x, y) = Printf.sprintf "%d,%d" x y in
  let visited = Hash_set.create (module String) in
  Hash_set.add visited (to_key start);
  let rec go queue splitters =
    match queue with
    | [] -> splitters
    | current :: rest ->
        let ns = neighbors grid current in
        let s' = if List.length ns = 2 then splitters + 1 else splitters in
        let new_ns =
          List.filter ns ~f:(fun n -> not (Hash_set.mem visited (to_key n)))
        in
        List.iter new_ns ~f:(fun n -> Hash_set.add visited (to_key n));
        go (rest @ new_ns) s'
  in
  go [ start ] 0

let count_paths grid start =
  let to_key (x, y) = Printf.sprintf "%d,%d" x y in
  let memo = Hashtbl.create (module String) in
  let rec dfs pos =
    let key = to_key pos in
    match Hashtbl.find memo key with
    | Some count -> count
    | None ->
        let ns = neighbors grid pos in
        let count =
          match ns with [] -> 1 | _ -> List.sum (module Int) ns ~f:dfs
        in
        Hashtbl.set memo ~key ~data:count;
        count
  in
  dfs start

let part1 lines =
  let grid = lines |> List.map ~f:String.to_array |> List.to_array in
  let start_x =
    Array.findi grid.(0) ~f:(fun _ c -> Char.equal c 'S')
    |> Option.value_exn |> fst
  in
  bfs grid (start_x, 0)

let part2 lines =
  let grid = lines |> List.map ~f:String.to_array |> List.to_array in
  let start_x =
    Array.findi grid.(0) ~f:(fun _ c -> Char.equal c 'S')
    |> Option.value_exn |> fst
  in
  count_paths grid (start_x, 0)

let () =
  let lines = read_input () in
  printf "Part 1: %d\n" (part1 lines);
  printf "Part 2: %d\n" (part2 lines)

File: 07/solution.ts
const input = await Bun.file(`${import.meta.dir}/input.txt`).text();
const lines = input.trim().split("\n");

type Point = { x: number; y: number };

function neighbors(grid: string[][], { x, y }: Point) {
	const cell = grid[y][x];
	const deltaMap: Record<string, [number, number][]> = {
		S: [[0, 1]],
		".": [[0, 1]],
		"^": [
			[1, 1],
			[-1, 1],
		],
	};
	const deltas = deltaMap[cell] ?? [];

	return deltas
		.map(([dx, dy]) => ({ x: x + dx, y: y + dy }))
		.filter(
			(n) => n.x >= 0 && n.x < grid[0].length && n.y >= 0 && n.y < grid.length,
		);
}

const toKey = (p: Point) => `(${p.x},${p.y})`;

function bfs(grid: string[][], start: Point) {
	const splitters = new Set<string>();
	const queue: Point[] = [start];
	const visited = new Set<string>();
	visited.add(toKey(start));

	while (queue.length > 0) {
		// biome-ignore lint/style/noNonNullAssertion: always guaranteed to exist
		const current = queue.shift()!;
		const ns = neighbors(grid, current);
		if (ns.length === 2) {
			splitters.add(toKey(current));
		}

		for (const neighbor of ns) {
			if (!visited.has(toKey(neighbor))) {
				visited.add(toKey(neighbor));
				queue.push(neighbor);
			}
		}
	}

	return splitters.size;
}

function countPaths(grid: string[][], start: Point) {
	const memo = new Map<string, number>();

	function dfs(pos: Point): number {
		const key = toKey(pos);
		// biome-ignore lint/style/noNonNullAssertion: we just check it has it
		if (memo.has(key)) return memo.get(key)!;

		const ns = neighbors(grid, pos);
		if (ns.length === 0) return 1;

		const count = ns.reduce((sum, n) => sum + dfs(n), 0);
		memo.set(key, count);
		return count;
	}

	return dfs(start);
}

function part1() {
	const grid = lines.map((l) => l.split(""));
	const start = { x: grid[0].findIndex((p) => p === "S"), y: 0 };
	return bfs(grid, start);
}

function part2() {
	const grid = lines.map((l) => l.split(""));
	const start = { x: grid[0].findIndex((p) => p === "S"), y: 0 };
	return countPaths(grid, start);
}

console.log("Part 1:", part1());
console.log("Part 2:", part2());
